### PR TITLE
Set a fixed upper limit for stack promotion in withUnsafeTemporaryAllocation() beyond which no stack allocation could ever succeed

### PR DESCRIPTION
**Pull request by grynspan**
Set a fixed upper limit for stack promotion in withUnsafeTemporaryAllocation() beyond which no stack allocation could ever succeed.

This change improves compile-time optimization of `withUnsafeTemporaryAllocation()` when passed a buffer size that is known at compile-time and which is too large to ever fit on the stack.

This change does _not_ mean that all allocations smaller than `upperBound` are stack-allocated: it means that such allocations are passed to `swift_stdlib_isStackAllocationSafe()` for potential further checking, but allocations _larger_ than `upperBound` always go on the heap (which they would _anyway_, so the function call is not necessary.)

So in short, this change eliminates an unnecessary function call for very large temporary allocations that always end up on the heap.

### CODE REVIEW

**Comment by grynspan** [comment]
@swift-ci please test

File: stdlib/public/core/TemporaryAllocation.swift [hunk 1]
```diff
@@ -84,10 +84,23 @@ internal func _isStackAllocationSafe(byteCount: Int, alignment: Int) -> Bool {
...
   // stack-promoted objects.
-  if byteCount <= 1024 {
+  if byteCount <= 1024 /* 1KiB */ {
     return true
   }
 
+  // Allocations larger than this limit are never candidates for stack promotion
+  // and will always be placed on the heap. If you're working on a platform with
+  // a particularly constrained stack limit, you might want to special-case this
+  // test and use a different upper bound.
```
Comment (gottesmm) [inline stdlib/public/core/TemporaryAllocation.swift:94]: Can you add a TODO that says we should find a way to make this assert if a new platform is added and this isn't updated.

Also I haven't thought about how to do it, but it would be great to have a test.

Comment (grynspan) [reply stdlib/public/core/TemporaryAllocation.swift:94]: I don't think there's anything to test for!

Comment (grynspan) [reply stdlib/public/core/TemporaryAllocation.swift:94]: Added a TODO and an additional test to show that the emitted code just heap-allocates (haven't actually, you know, tested it though.)

**Comment by grynspan** [comment]
@swift-ci please test

File: test/IRGen/temporary_allocation/codegen.swift [hunk 2]
```diff
@@ -66,32 +66,42 @@ withUnsafeTemporaryAllocation(of: Void.self, capacity: 2) { buffer in
...
-// CHECK: [[STACK_PTR:%[0-9]+]] = ptrtoint [268435455 x i8]* [[STACK_PTR_RAW]] to [[WORD]]
+// CHECK: [[STACK_PTR_RAW:%temp_alloc[0-9]*]] = alloca [65535 x i8], align 1
+// CHECK: [[STACK_PTR:%[0-9]+]] = ptrtoint [65535 x i8]* [[STACK_PTR_RAW]] to [[WORD]]
 // CHECK: call swiftcc void @blackHole([[WORD]] [[STACK_PTR]])
 // CHECK: call void @llvm.stackrestore(i8* [[SPSAVE]])
 
+// MARK: Excessively large allocation
+
+// An excessively large allocation should always end up on the heap.
+withUnsafeTemporaryAllocation(byteCount: 0xFFFF_FFFF, alignment: 1) { buffer in
```
Comment (karwa) [inline test/IRGen/temporary_allocation/codegen.swift:101]: Isn't the idea that this should _never_ call `swift_stdlib_isStackAllocationSafe`? Perhaps it's worth checking that.

Comment (karwa) [reply test/IRGen/temporary_allocation/codegen.swift:101]: Right, that's what I said. Shouldn't this test include a `CHECK-NOT` for the call to `swift_stdlib_isStackAllocationSafe`?

Comment (grynspan) [reply test/IRGen/temporary_allocation/codegen.swift:101]: Ah! Sure, I can add that. :)
